### PR TITLE
Add pre/post transfer balance checks

### PR DESCRIPTION
**Pull request by asoong**
- Pre/post transfer balance checks.

### CODE REVIEW

File: contracts/core/TransferProxy.sol [hunk 1]
```diff
@@ -76,11 +80,18 @@ contract TransferProxy is
         external
         onlyAuthorized
     {
+        // Retrive current balance of token for the vault
```
Comment (felix2feng) [inline contracts/core/TransferProxy.sol]: "Retrieve"

File: contracts/core/TransferProxy.sol [hunk 2]
```diff
@@ -76,11 +80,18 @@ contract TransferProxy is
...
+        uint exstingVaultBalance = ERC20(_tokenAddress).balanceOf(vaultAddress);
```
Comment (felix2feng) [inline contracts/core/TransferProxy.sol]: "existingVaultBalance" vs "exstingVaultBalance"?

Comment (felix2feng) [reply contracts/core/TransferProxy.sol]: This works for now. Though, did you have any thoughts re: using something like dYdX's TokenInteract?

File: contracts/core/Vault.sol [hunk 3]
```diff
@@ -92,11 +92,18 @@ contract Vault is
         isNonZero(_quantity)
         isValidDestination(_to)
     {
+        // Retrive current balance of token for the vault
```
Comment (felix2feng) [inline contracts/core/Vault.sol]: Same as above

File: contracts/core/Vault.sol [hunk 4]
```diff
@@ -92,11 +92,18 @@ contract Vault is
...
+        uint exstingVaultBalance = ERC20(_tokenAddress).balanceOf(this);
```
Comment (felix2feng) [inline contracts/core/Vault.sol]: Same as above

File: test/transferProxy.spec.ts [hunk 5]
```diff
@@ -155,16 +178,19 @@ contract("TransferProxy", (accounts) => {
...
+      await approveTransfer(mockToken, transferProxy.address, approver);
     });
 
     // Subject
-    const amountToTransfer = STANDARD_INITIAL_TOKENS;
+    const amountToTransfer: BigNumber = STANDARD_INITIAL_TOKENS;
+    let tokenAddress: Address;
 
     async function subject(): Promise<string> {
+      const tokenToTransfer = tokenAddress || mockToken.address;
```
Comment (felix2feng) [inline test/transferProxy.spec.ts]: Maybe add a comment?

File: test/vault.spec.ts [hunk 6]
```diff
@@ -113,6 +108,26 @@ contract("Vault", (accounts) => {
       );
   };
 
+  async function deployTokenWithFee(fee: BigNumber, initialAccount: Address, from: Address = ownerAccount) {
```
Comment (felix2feng) [inline test/vault.spec.ts]: I wonder if there's a way to make this a library so there's no duplication across `vault.spec` and `transferProxy.spec`

Comment (asoong) [reply test/vault.spec.ts]: For sure, we will DRY up testing shortly.

**Pull request merged**